base_trgm_search/models/trgm_index.py

```diff
@@ -0,0 +1,224 @@
+        column_name = self.field_id.name
+        index_type = self.index_type
+        index_name = '%s_%s_idx' % (column_name, index_type)
+        index_exists, index_name = self.get_not_used_index(
+            index_name, table_name)
+
+        if not index_exists:
+            self.env.cr.execute("""
+            CREATE INDEX %(index)s
+            ON %(table)s
+            USING %(indextype)s (%(column)s %(indextype)s_trgm_ops);
+            """, {
+                'table': AsIs(table_name),
+                'index': AsIs(index_name),
+                'column': AsIs(column_name),
+                'indextype': AsIs(index_type)
+            })
+        return index_name
+
+    @api.model
+    def index_exists(self, model_name, field_name):
+        field = self.env['ir.model.fields'].search([
+            ('model', '=', model_name), ('name', '=', field_name)], limit=1)
+
+        if not field:
+            return False
+
+        trgm_index = self.search([('field_id', '=', field.id)], limit=1)
+        return bool(trgm_index)
+
+    @api.model
+    def create(self, vals):
+        rec = super(TrgmIndex, self).create(vals)
+        rec.index_name = rec.create_index()
+        return rec
+
+    @api.multi
+    def unlink(self):
+        for rec in self:
+            self.env.cr.execute("""
+                DROP INDEX IF EXISTS %(index)s;
+                """, {
+                'index': AsIs(rec.index_name),
+            })
+        return super(TrgmIndex, self).unlink()
+
+
+def patch_leaf_trgm(method):
+    def decorate_leaf_to_sql(self, eleaf):
+        model = eleaf.model
+        leaf = eleaf.leaf
+        left, operator, right = leaf
+        table_alias = '"%s"' % (eleaf.generate_alias())
+
+        if operator == '%':
+            sql_operator = '%%'
+            params = []
+
+            if left in model._columns:
+                format = model._columns[left]._symbol_set[0]
+                column = '%s.%s' % (table_alias, expression._quote(left))
+                query = '(%s %s %s)' % (column, sql_operator, format)
+            elif left in expression.MAGIC_COLUMNS:
+                query = "(%s.\"%s\" %s %%s)" % (
+                    table_alias, left, sql_operator)
+                params = right
+            else:  # Must not happen
+                raise ValueError(
+                    "Invalid field %r in domain term %r" % (left, leaf))
+
+            if left in model._columns:
+                params = model._columns[left]._symbol_set[1](right)
+
+            if isinstance(params, basestring):
+                params = [params]
+            return query, params
+        elif operator == 'inselect':
+            right = (right[0].replace(' % ', ' %% '), right[1])
+            eleaf.leaf = (left, operator, right)
+        return method(self, eleaf)
+    return decorate_leaf_to_sql
+
+expression.expression._expression__leaf_to_sql = patch_leaf_trgm(
+    expression.expression._expression__leaf_to_sql)
```

Comment:
Please perform patching in register_hook to prevent patching when the module is not installed on any database. And tests, yes!